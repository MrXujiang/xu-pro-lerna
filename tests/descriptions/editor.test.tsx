import React, { useRef } from 'react';
import { InputNumber, Form } from 'antd';
import type { RowEditableConfig } from '@ant-design/pro-utils';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import type {
  ProDescriptionsActionType,
  ProDescriptionsItemProps,
} from '@ant-design/pro-descriptions';
import Descriptions from '@ant-design/pro-descriptions';
import { mount, render } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { waitForComponentToPaint } from '../util';

type DataSourceType = {
  id: number;
  title?: string;
  labels?: {
    name: string;
    color: string;
  }[];
  state?: string;
  time?: {
    created_at?: string;
  };
  children?: DataSourceType;
};

const defaultData: DataSourceType = {
  id: 624748504,
  title: 'üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô',
  labels: [{ name: 'bug', color: 'error' }],
  time: {
    created_at: '2020-05-26T09:42:56Z',
  },
  state: 'processing',
};

const columns: ProDescriptionsItemProps<DataSourceType>[] = [
  {
    dataIndex: 'index',
    valueType: 'indexBorder',
    renderFormItem: () => <InputNumber />,
  },
  {
    title: 'Ê†áÈ¢ò',
    dataIndex: 'title',
    copyable: true,
    ellipsis: true,
    tip: 'Ê†áÈ¢òËøáÈïø‰ºöËá™Âä®Êî∂Áº©',
    formItemProps: {
      rules: [
        {
          required: true,
          message: 'Ê≠§È°π‰∏∫ÂøÖÂ°´È°π',
        },
      ],
    },
  },
  {
    title: 'Áä∂ÊÄÅ',
    dataIndex: 'state',
    valueType: 'select',
    valueEnum: {
      all: { text: 'ÂÖ®ÈÉ®', status: 'Default' },
      open: {
        text: 'Êú™Ëß£ÂÜ≥',
        status: 'Error',
      },
      closed: {
        text: 'Â∑≤Ëß£ÂÜ≥',
        status: 'Success',
      },
      processing: {
        text: 'Ëß£ÂÜ≥‰∏≠',
        status: 'Processing',
      },
    },
  },
  {
    title: 'ÂàõÂª∫Êó∂Èó¥',
    dataIndex: ['time', 'created_at'],
    valueType: 'date',
  },
];

const DescriptionsDemo = (
  props: {
    type?: 'multiple';
    defaultKeys?: React.Key[];
    editorRowKeys?: React.Key[];
    onEditorChange?: (editorRowKeys: React.Key[]) => void;
    dataSource?: DataSourceType;
    onDataSourceChange?: (dataSource: DataSourceType) => void;
  } & RowEditableConfig<DataSourceType>,
) => {
  const [form] = Form.useForm();
  const actionRef = useRef<ProDescriptionsActionType>();
  const [editableKeys, setEditorRowKeys] = useMergedState<React.Key[]>(
    () => props.defaultKeys || [],
    {
      value: props.editorRowKeys,
      onChange: props.onEditorChange,
    },
  );
  const [dataSource, setDataSource] = useMergedState<DataSourceType, DataSourceType>(
    props.dataSource as any,
    {
      value: props.dataSource,
      onChange: props.onDataSourceChange,
    },
  );
  return (
    <Descriptions<DataSourceType>
      columns={columns}
      actionRef={actionRef}
      request={async () => ({
        data: defaultData,
        total: 3,
        success: true,
      })}
      title={
        <a
          id="reset_test"
          onClick={() => {
            form.resetFields();
          }}
        >
          ÈáçÁΩÆ
        </a>
      }
      dataSource={dataSource}
      onDataSourceChange={setDataSource}
      editable={{
        ...props,
        form,
        type: props.type,
        editableKeys,
        onSave: props.onSave,
        onChange: setEditorRowKeys,
      }}
    />
  );
};

describe('Descriptions', () => {
  it('üìù Descriptions close editable', async () => {
    const wrapper = mount(
      <Descriptions<DataSourceType> columns={columns} dataSource={defaultData} />,
    );
    await waitForComponentToPaint(wrapper, 100);
    expect(wrapper.find('ProForm').exists()).toBeFalsy();
  });

  it('üìù Descriptions support editable', async () => {
    const wrapper = mount(
      <Descriptions<DataSourceType> columns={columns} dataSource={defaultData} editable={{}} />,
    );
    await waitForComponentToPaint(wrapper, 100);
    expect(wrapper.find('ProForm').exists()).toBeTruthy();
  });

  it('üìù support onEditorChange', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <DescriptionsDemo
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(0).simulate('click');
    });
    await waitForComponentToPaint(wrapper);
    expect(fn).toBeCalledWith(['title']);
  });

  it('üìù support set Form', async () => {
    const wrapper = mount(<DescriptionsDemo editorRowKeys={['title']} />);
    await waitForComponentToPaint(wrapper, 1000);

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`.ant-input`)
        .simulate('change', {
          target: {
            value: 'test',
          },
        });
    });
    await waitForComponentToPaint(wrapper, 200);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`.ant-input`)
        .props().value,
    ).toBe('test');

    act(() => {
      wrapper.find('#reset_test').simulate('click');
    });
    await waitForComponentToPaint(wrapper, 200);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`.ant-input`)
        .props().value,
    ).toBe('üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô');
  });

  it('üìù renderFormItem run defaultRender', async () => {
    const wrapper = render(
      <Descriptions<DataSourceType>
        editable={{
          editableKeys: ['title'],
        }}
        columns={[
          {
            dataIndex: 'title',
            renderFormItem: (item, config) => {
              return config.defaultRender(item);
            },
          },
        ]}
        dataSource={defaultData}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('üìù columns support editable test', async () => {
    const wrapper = render(
      <Descriptions
        editable={{
          editableKeys: ['title'],
        }}
        columns={[
          {
            dataIndex: 'title',
            editable: (text, record, index) => {
              return index === 1;
            },
          },
          {
            dataIndex: 'title2',
            editable: false,
          },
        ]}
        dataSource={defaultData}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('üìù support actionRender', async () => {
    const wrapper = render(
      <Descriptions
        editable={{
          editableKeys: ['title'],
          actionRender: () => [
            <div key="test" id="test">
              xx
            </div>,
          ],
        }}
        columns={[
          {
            dataIndex: 'title',
            editable: (text, record, index) => {
              return index === 1;
            },
          },
          {
            dataIndex: 'title2',
            editable: false,
          },
        ]}
        dataSource={defaultData}
      />,
    );
    expect(wrapper.find('div#test').text()).toBe('xx');
  });

  it('üìù support editorRowKeys', async () => {
    const wrapper = mount(<DescriptionsDemo editorRowKeys={['title']} />);
    await waitForComponentToPaint(wrapper, 1000);
    // Á¨¨‰∏ÄË°åÂ∫îËØ•ÁºñËæëÊÄÅ
    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find('input')
        .exists(),
    ).toBeTruthy();

    // Á¨¨‰∫åË°å‰∏çÂ∫îËØ•ÊòØÁºñËæëÊÄÅ
    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find('input')
        .exists(),
    ).toBeFalsy();
  });

  it('üìù support cancel click', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <DescriptionsDemo
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(0).simulate('click');
    });
    await waitForComponentToPaint(wrapper, 1000);
    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find('input')
        .exists(),
    ).toBeTruthy();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`span.anticon-close`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 1000);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find('input')
        .exists(),
    ).toBeFalsy();
  });

  it('üìù support cancel click render false', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <DescriptionsDemo
        onEditorChange={(keys) => {
          fn(keys);
        }}
        onCancel={async () => false}
      />,
    );
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(0).simulate('click');
    });
    await waitForComponentToPaint(wrapper, 1000);
    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find('input')
        .exists(),
    ).toBeTruthy();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`span.anticon-close`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 1000);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find('input')
        .exists(),
    ).toBeFalsy();
  });

  it('üìù type=single, only edit one rows', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <DescriptionsDemo
        defaultKeys={['state']}
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(0).simulate('click');
    });

    await waitForComponentToPaint(wrapper, 1000);

    expect(fn).not.toBeCalled();
  });

  it('üìù type=multiple, edit multiple rows', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <DescriptionsDemo
        type="multiple"
        defaultKeys={['state']}
        onEditorChange={(keys) => {
          fn(keys);
        }}
      />,
    );
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(0).simulate('click');
    });
    await waitForComponentToPaint(wrapper, 1000);
    expect(fn).toBeCalledWith(['state', 'title']);
  });

  it('üìù support onSave', async () => {
    const fn = jest.fn();
    const wrapper = mount(<DescriptionsDemo onSave={(key) => fn(key)} />);
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(1).simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find('input')
        .exists(),
    ).toBeTruthy();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find(`span.anticon-check`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(fn).toBeCalledWith('state');
  });

  it('üìù support onSave support false', async () => {
    const fn = jest.fn();
    const wrapper = mount(
      <DescriptionsDemo
        onSave={async (key) => {
          fn(key);
          return false;
        }}
      />,
    );
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(1).simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find('input')
        .exists(),
    ).toBeTruthy();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find(`span.anticon-check`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find('input')
        .exists(),
    ).toBeTruthy();

    expect(fn).toBeCalledWith('state');
  });

  it('üìù support onCancel', async () => {
    const fn = jest.fn();
    const wrapper = mount(<DescriptionsDemo onCancel={(key) => fn(key)} />);
    await waitForComponentToPaint(wrapper, 1000);
    act(() => {
      wrapper.find('span.anticon-edit').at(1).simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find('input')
        .exists(),
    ).toBeTruthy();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(1)
        .find(`span.anticon-close`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(fn).toBeCalledWith('state');
  });

  it('üìù support form rules', async () => {
    const fn = jest.fn();
    const wrapper = mount(<DescriptionsDemo onSave={(key, row) => fn(row.title)} />);
    await waitForComponentToPaint(wrapper, 1000);

    act(() => {
      wrapper.find('span.anticon-edit').at(0).simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);
    expect(
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find('input')
        .exists(),
    ).toBeTruthy();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`.ant-input`)
        .simulate('change', {
          target: {
            value: '',
          },
        });
    });
    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`span.anticon-check`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    // Ê≤°ÊúâÈÄöËøáÈ™åËØÅÔºå‰∏çËß¶Âèë onSave
    expect(fn).not.toBeCalled();

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`input.ant-input`)
        .simulate('change', {
          target: {
            value: 'qixian',
          },
        });
    });

    act(() => {
      wrapper
        .find('td.ant-descriptions-item .ant-descriptions-item-content')
        .at(0)
        .find(`span.anticon-check`)
        .simulate('click');
    });

    await waitForComponentToPaint(wrapper, 200);

    expect(fn).toBeCalledWith('qixian');
  });
});
