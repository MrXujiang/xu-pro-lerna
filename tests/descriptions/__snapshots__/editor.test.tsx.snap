// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Descriptions üìù columns support editable test 1`] = `
<div
  class="ant-descriptions ant-pro-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                <div
                  style="margin-top:-5px;margin-bottom:-5px;margin-left:0;margin-right:0"
                >
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right:8px"
                    >
                      <div
                        class="ant-row ant-form-item"
                        style="margin-top:-5px;margin-bottom:-5px;margin-left:0;margin-right:0;margin:0"
                      >
                        <div
                          class="ant-col ant-form-item-control"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <span
                                class="ant-input-affix-wrapper"
                              >
                                <input
                                  class="ant-input"
                                  placeholder="ËØ∑ËæìÂÖ•"
                                  type="text"
                                  value="üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô"
                                />
                                <span
                                  class="ant-input-suffix"
                                >
                                  <span
                                    aria-label="close-circle"
                                    class="anticon anticon-close-circle ant-input-clear-icon"
                                    role="button"
                                    tabindex="-1"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="close-circle"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right:8px"
                    >
                      <a>
                        <span
                          aria-label="check"
                          class="anticon anticon-check"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="check"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                            />
                          </svg>
                        </span>
                      </a>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right:8px"
                    />
                    <div
                      class="ant-space-item"
                    >
                      <a>
                        <span
                          aria-label="close"
                          class="anticon anticon-close"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                            />
                          </svg>
                        </span>
                      </a>
                    </div>
                  </div>
                </div>
              </span>
            </div>
          </td>
          <td
            class="ant-descriptions-item"
            colspan="2"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                -
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions üìù renderFormItem run defaultRender 1`] = `
<div
  class="ant-descriptions ant-pro-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="3"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-content"
              >
                <div
                  style="margin-top:-5px;margin-bottom:-5px;margin-left:0;margin-right:0"
                >
                  <div
                    class="ant-space ant-space-horizontal ant-space-align-center"
                  >
                    <div
                      class="ant-space-item"
                      style="margin-right:8px"
                    >
                      <div
                        class="ant-row ant-form-item"
                        style="margin-top:-5px;margin-bottom:-5px;margin-left:0;margin-right:0;margin:0"
                      >
                        <div
                          class="ant-col ant-form-item-control"
                        >
                          <div
                            class="ant-form-item-control-input"
                          >
                            <div
                              class="ant-form-item-control-input-content"
                            >
                              <span
                                class="ant-input-affix-wrapper"
                              >
                                <input
                                  class="ant-input"
                                  placeholder="ËØ∑ËæìÂÖ•"
                                  type="text"
                                  value="üêõ [BUG]yarn installÂëΩ‰ª§ antd2.4.5‰ºöÊä•Èîô"
                                />
                                <span
                                  class="ant-input-suffix"
                                >
                                  <span
                                    aria-label="close-circle"
                                    class="anticon anticon-close-circle ant-input-clear-icon"
                                    role="button"
                                    tabindex="-1"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="close-circle"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right:8px"
                    >
                      <a>
                        <span
                          aria-label="check"
                          class="anticon anticon-check"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="check"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z"
                            />
                          </svg>
                        </span>
                      </a>
                    </div>
                    <div
                      class="ant-space-item"
                      style="margin-right:8px"
                    />
                    <div
                      class="ant-space-item"
                    >
                      <a>
                        <span
                          aria-label="close"
                          class="anticon anticon-close"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                            />
                          </svg>
                        </span>
                      </a>
                    </div>
                  </div>
                </div>
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
